dbms/src/Flash/Planner/Plans/PhysicalTopN.cpp

```diff
@@ -72,11 +71,22 @@ void PhysicalTopN::buildPipelineExecGroup(
     PipelineExecutorStatus & exec_status,
     PipelineExecGroupBuilder & group_builder,
     Context & context,
-    size_t /*concurrency*/)
+    size_t concurrency)
 {
     executeExpression(exec_status, group_builder, before_sort_actions, log);
 
-    executeLocalSort(exec_status, group_builder, order_descr, limit, context, log);
+    // If the `limit` is very large, using a `final sort` can avoid outputting excessively large amounts of data.
+    // TODO find a suitable threshold is necessary; 10000 is just a value picked without much consideration.
+    if (group_builder.concurrency() * limit <= 10000)
```

Comment:
why multiply limit with concurrency?

Reply:
Because `localSort` outputs `limit` rows for each thread, the total output is `concurrency * limit`.